Introduce LogicalPlan invariants, begin automatically checking them

wiedld opened the pull request:
## Which issue does this PR close?

Part of #13652 

## Rationale for this change

The [original discussion](https://github.com/apache/datafusion/issues/13525) included implicit changes which can cause problems when trying to upgrade DF. One class of bugs are related to user-constructed LPs, and the mutations of these LPs. This PR is a first step to programmatically enforce the rules of what should, and should not, be done.


## What changes are included in this PR?


* defined what must always be true for an LP.
    * This applies to all LPs, including those constructed by user (e.g. own language frontends). 
* define the contract for semantic mutation.
    * The LP may not be executable before the Analyzer passes.
    * The LP must be executable after the Analyzer passes.
* define what further can be mutated during the LP optimizer passes.
    * The LP must be executable before, and after, the optimizer passes.
    * The LP schema cannot be mutated by the optimizer.

## Are these changes tested?

By existing tests, and are benchmarked for impact to planning time.

## Are there any user-facing changes?

There is a new `LogicalPlan::check_invariants` public api.

wiedld commented on datafusion/optimizer/src/analyzer/mod.rs:
```diff
@@ -137,6 +137,19 @@ impl Analyzer {
     where
         F: FnMut(&LogicalPlan, &dyn AnalyzerRule),
     {
+        /* current test failure:
+           External error: statement is expected to fail with error:
+                   (regex) DataFusion error: check_analyzed_plan\ncaused by\nError during planning: Scalar subquery should only return one column, but found 2: t2.t2_id, t2.t2_name
+               but got error:
+                   DataFusion error: Error during planning: Scalar subquery should only return one column, but found 2: t2.t2_id, t2.t2_name
+               [SQL] SELECT t1_id, t1_name, t1_int, (select t2_id, t2_name FROM t2 WHERE t2.t2_id = t1.t1_int) FROM t1
+               at test_files/subquery.slt:436
+        */
+        // verify at the start, before the first LP analyzer pass.
+        // check_plan(&plan).map_err(|e| {
+        //     DataFusionError::Context("check_plan_before_analyzers".to_string(), Box::new(e))
+        // })?;
```
Note this existing test failure.

TODO (if we proceed): uncomment check and debug test.

wiedld commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -451,6 +468,33 @@ impl Optimizer {
+    // verify invariant: fields must have unique names
+    assert_unique_field_names(plan)?;
+
+    /* This current fails for:
+       - execution::context::tests::cross_catalog_access
+       - at test_files/string/string.slt:46
+               External error: query failed: DataFusion error: Optimizer rule 'eliminate_nested_union' failed
+    */
+    // verify invariant: equivalent schema across union inputs
+    // assert_unions_are_valid(check_name, plan)?;
```
Note this existing test failure.

TODO (if we proceed): uncomment check and debug test.

alamb replied:
I think this function might be better as a method of LogicalPlan. Maybe something like

```rust
LogicalPlan::check_invariants()
```

Or something

Sl1mb0 replied:
My half a cent:
```rust
LogicalPlan::assert_invariants()
```

alamb reviewed (commented):
Thank you @wiedld  -- I really like this idea

I filed an issue for this idea here: https://github.com/apache/datafusion/issues/13652

@findepi  @jonahgao or @Omega359  I wonder if you have any thoughts on this basic idea?

Omega359 commented:
I love the idea of having additional testing like this! It's definitely out of my area of expertise but a few questions:

- Should invariants be pluggable? For example, Influxdb might have additional invariants that another derivative of DF does not (or vice versa)? 
- Is the invariant testing overhead acceptable for production or should it there be the capability to disable it (enabled by default of course!) ? 
- Is there an equivalent set of invariants elsewhere in the core besides the LP (maybe physical plan? Just guessing here as I'm definitely no expert here) where this approach could be duplicated ?

wiedld commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -451,6 +468,33 @@ impl Optimizer {
+    // verify invariant: equivalent schema across union inputs
+    // assert_unions_are_valid(check_name, plan)?;
+
+    // TODO: trait API and provide extension on the Optimizer to define own validations?
```
Here is a mention of the extensibility of invariants. Options include:
* for general invariants:
   * defined as being checked before/after each OptimizerRule, and applied here in `check_plan()` (or equivalent code)
   * we could provide `Optimizer.invariants = Vec<Arc<dyn InvariantCheck>>` for user-defined invariants
* for invariants specific for a given OptimizerRule:
   * we could provide `OptimizerRule::check_invariants()` such that certain invariants are only checked for a given rule (instead of all rules)
   * for a user-defined OptimizerRule, users can also check their own invariants

Ditto for the AnalyzerRule passes. Altho I wasn't sure about how much is added complexity and planning time overhead - as @Omega359 mentions we could [make it configurable](https://github.com/apache/datafusion/pull/13651#issuecomment-2519006923) (e.g. run for CI and debugging in downstream projects). 

This WIP is about proposing different ideas of what we could do. 🤔

jonahgao replied:
Maybe it can be controlled through environment variables, similar to `RUST_LOG`. Enable it for debugging when problems are encountered or during an upgrade.

findepi replied:
do you have example use-case for user-defined plan invariants?

alamb replied:
We have some special invariants for our `SortPreservingMerge` replacement, `ProgressiveEval` (related to time ranges of parquet files) that would be great to be able to encode

alamb replied:
> Maybe it can be controlled through environment variables, similar to RUST_LOG or RUST_BACKTRACE. Enable it for debugging when problems are encountered or during an upgrade.

We could also add it as a `debug_assert!` after each optimizer pass and call the real validation
1. After analyze
2. After all the optimizer passes

findepi replied:
> We have some special invariants for our `SortPreservingMerge` replacement, `ProgressiveEval` (related to time ranges of parquet files) that would be great to be able to encode

is this about `LogicalPlan::Extension`? I agree it makes sense to support validation of these if we validate the overall plan.

wiedld replied:
I've added this to the followup items list: https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777

findepi commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -476,6 +520,38 @@ pub(crate) fn assert_schema_is_the_same(
+/// This invariant is subject to change.
+/// refer: <https://github.com/apache/datafusion/issues/13525#issuecomment-2494046463>
+fn assert_unique_field_names(plan: &LogicalPlan) -> Result<()> {
+    plan.schema().check_names()?;
```
Is `check_names` also called whenever creating new DFSchema?

wiedld replied:
Yes, on every creation. But not after [every merge](https://github.com/apache/datafusion/blob/bd2c975df88c9e95f2f6b5f2abed5638904e2ac0/datafusion/common/src/dfschema.rs#L294). Which should be ok if no bug is introduced in the merge -- altho I would prefer to add the check there.

findepi commented:
I like the verification. The only concern is the overhead of doing it, especially for large plans. Every optimizer pass gets a new verification of all plan nodes. i hope one day we go towards more iterative optimization and then we should be able to verify plan invariants locally. Which would guarantee that the verification overhead is linearly proportional to number of modifications applied to the plan. This would require eg that get_type is a constant operation (https://github.com/apache/datafusion/issues/12604), since plan verification should also include that the types do match.

alamb commented:
I think it would also be great if we could consider moving the invariant check directly into `LogicalPlan` -- for exmaple
`LogicalPlan::check_invariants`

This would make this function easier to discover and for others to understand what the invariants are.

wiedld commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1125,6 +1126,13 @@ impl LogicalPlan {
+    /// These are invariants to hold true for each logical plan.
+    pub fn assert_invariants(&self) -> Result<()> {
+        assert_unique_field_names(self)?;
+
```
I removed the `assert_unions_are_valid` because I was not sure if it should be an invariant, and it caused ~17 test failures.

wiedld commented on datafusion/optimizer/src/analyzer/mod.rs:
```diff
@@ -164,18 +172,38 @@ impl Analyzer {
+fn assert_valid_semantic_plan(plan: &LogicalPlan) -> Result<()> {
+    plan.assert_invariants()?;
+
+    // TODO: should this be moved to LogicalPlan::assert_invariants?
+    assert_subqueries_are_valid(plan)?;
```
The `assert_subqueries_are_valid()` runs [these checks](https://github.com/apache/datafusion/blob/bd2c975df88c9e95f2f6b5f2abed5638904e2ac0/datafusion/optimizer/src/analyzer/subquery.rs#L28-L34). The existing code only performs these checks during after all of the analyzer passes have completed. If I run these checks before the first analyzer pass => we get test failures.

My next steps will be to figure out which of these test failures are due to code bugs.

wiedld replied:
Also note that what makes a valid plan (plan checks) are different from [invariants](https://datafusion.apache.org/contributor-guide/specification/invariants.html#id4). Which is why I have this plan check here (and not under `LogicalPlan::assert_invariants`). But I'm not sure.

alamb replied:
> If I run these checks before the first analyzer pass => we get test failures.

I think this is somewhat "expected" -- part of the Analyzer's job is to make LogicalPlan's more "valid" as described here https://docs.rs/datafusion/latest/datafusion/optimizer/trait.AnalyzerRule.html

Perhaps we can make this part of `LogicalPlan::assert_invariants`, with an enum that explains the rules being followed? 

```rust
enum InvariantLevel {
  /// Invariants that are always true in DataFusion `LogicalPlan`s
  /// such as the number of expected children and no duplicated output fields
  Always, 
  /// Invariants that must hold true for the plan to be "executable"
  /// such as the type and number of function arguments are correct and
  /// that wildcards have been expanded
  ///
  /// To ensure a LogicalPlan satisfies the `Executable` inariants, run the
  /// `Analyzer`
  Executable,
}

impl LogicalPlan {
  /// checks that the plan conforms to the listed invariant level, returning an Error if not
  fn check_invariants(&self, check: InvariantLevel) -> Result<()> {
  ...
  }
}
```

That way the `assert_subqueries_are_valid` check would always be available

wiedld commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1125,6 +1126,14 @@ impl LogicalPlan {
+    pub fn assert_invariants(&self) -> Result<()> {
+        // Refer to <https://datafusion.apache.org/contributor-guide/specification/invariants.html#relation-name-tuples-in-logical-fields-and-logical-columns-are-unique>
+        assert_unique_field_names(self)?;
+
```
I removed the `assert_unions_are_valid` because I was not sure if it should be a LP invariant (it's [not listed here](https://datafusion.apache.org/contributor-guide/specification/invariants.html#id4)), and it caused ~17 test failures.

alamb replied:
Good call -- I think we could/should discover/enforce new invariants as follow on PRs after we have sorted out this basic API

wiedld commented:
The updates focus on dividing out LP invariants, vs analyzer pass checks, vs optimizer pass checks (including invariants listed as the analyzer's responsibility). I added in refernces links to invariants in the docs, in an attempt to delineate an invariant vs a valid plan. (But I'm muddy on this boundary.)

I'll make it configurable after I get feedback of whether this^^. 🙏🏼

alamb commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1125,6 +1126,14 @@ impl LogicalPlan {
         }
     }
 
+    /// These are invariants to hold true for each logical plan.
```
See suggestion below

alamb commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1125,6 +1126,14 @@ impl LogicalPlan {
 
+    /// These are invariants to hold true for each logical plan.
+    pub fn assert_invariants(&self) -> Result<()> {
+        // Refer to <https://datafusion.apache.org/contributor-guide/specification/invariants.html#relation-name-tuples-in-logical-fields-and-logical-columns-are-unique>
+        assert_unique_field_names(self)?;
```
I had forgotten about that documentation 🤔 

Maybe it would be a good thing to inline that information into the code rather than in a separate document that might be harder to find (not for you, I am just musing here for my own benefit)

wiedld replied:
I've added this to the followup items list: https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777

alamb commented on datafusion/expr/src/logical_plan/mod.rs:
```diff
@@ -20,6 +20,8 @@ mod ddl;
 pub mod display;
 pub mod dml;
 mod extension;
+pub(crate) mod invariants;
```
I didn't see a new `invariants` module in this PR 🤔

wiedld replied:
🤦🏼 . Fixed. TY.

alamb reviewed (commented):
This is looking great @wiedld -- thank you

I think the basic structure of assert_invariants is looking good. 

I had a suggestion for how to combine the two different invariant checks into a single API. 

In terms of planning:
1. Perhaps we can begin creating an epic for "Enforcing Invariants" and list items there to follow up on (like "Define the invariants for Union plans" for example)

In terms of this PR, once we solidify the API the only remaining concern I have is with potential performance slowdown of re-checking the invariants. When we have the code ready we can run the `sql_planning` benchmark to make sure this PR doesn't slow down planning. If it does, we can perhaps only run the invariant checks in debug builds

wiedld commented on datafusion/expr/src/logical_plan/invariants.rs:
```diff
@@ -109,12 +210,13 @@ pub fn check_subquery_expr(
             LogicalPlan::Projection(_)
             | LogicalPlan::Filter(_)
+            | LogicalPlan::TableScan(_)
             | LogicalPlan::Window(_)
             | LogicalPlan::Aggregate(_)
             | LogicalPlan::Join(_) => Ok(()),
             _ => plan_err!(
                 "In/Exist subquery can only be used in \
-                Projection, Filter, Window functions, Aggregate and Join plan nodes, \
+                Projection, Filter, TableScan, Window functions, Aggregate and Join plan nodes, \
```
The `push_down_filter` optimization pass can push a `IN(<subquery>)` into a TableScan. This change was required for [push_down_filter::tests::test_in_subquery_with_alias](https://github.com/apache/datafusion/blob/452a8f4fa8fbfd88a5a71310bc28b9f78685c66f/datafusion/optimizer/src/push_down_filter.rs#L2836) to pass.

wiedld commented:
Benchmarks in progress...

wiedld commented:
> (existing/main check): each optimizer pass only does a check that the schema is not changed

We could also reduce the time on this one by making it compile time with [Arc::ptr_eq](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.ptr_eq), but that means (a) a stricter contract (metadata & nullability cannot change) and (b) it requires code changes to chase down the current pointer_eq  failures. 🤔

alamb commented:
> I propose that we add another full InvariantLevel::Executable between each optimization pass, that only runs in debug mode (since that will be a performance impact as @findepi mentioned).

Perhaps we can do this as a follow on PR

> I can also make some of these check listed above into debug only, if the current performance change is unacceptable.

Yes, I think this is important for two reasons:
1.  I think we can't introduce a 25% planning performance regression
2. Adidng a check to help development that will generate errors in plans that used to run to completion will likely be seen as a regression  for users

I am checking out the rest of this PR now as well

alamb commented on datafusion/expr/src/logical_plan/invariants.rs:
```diff
@@ -281,13 +383,30 @@ fn check_mixed_out_refer_in_window(window: &Window) -> Result<()> {
     }
 }
 
+fn collect_subquery_cols(
```
this seems like a copy of the code in https://github.com/apache/datafusion/blob/344f0897d105cdd588d7a7ee8cef0646f844937a/datafusion/optimizer/src/utils.rs#L90-L89

It would be nice if it were shared -- perhaps it could be moved in https://github.com/apache/datafusion/blob/2ac8af894f9aefb3215b74044b7f177b42f7e65f/datafusion/expr/src/utils.rs#L18-L17 to be shared

alamb commented on datafusion/optimizer/src/analyzer/mod.rs:
```diff
@@ -46,17 +42,24 @@ pub mod expand_wildcard_rule;
-pub mod subquery;
 pub mod type_coercion;
 
+pub mod subquery {
```
👍

alamb commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -384,9 +394,16 @@ impl Optimizer {
                 .and_then(|tnr| {
-                    assert_schema_is_the_same(rule.name(), &starting_schema, &tnr.data)?;
+                    // verify after each optimizer pass.
+                    assert_valid_optimization(rule.name(), &tnr.data, &starting_schema)
```
To avoid performance regressions but still get the benefit of this check, I recommend changing this check to only run after each optimzer pass in debug mode (`#cfg(debug)`)

wiedld replied:
Done.

alamb commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -445,35 +462,38 @@ impl Optimizer {
             i += 1;
         }
+
+        // verify LP is valid, after the last optimizer pass.
```
💯  -- I think this is a very good thing to also check

alamb commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -445,35 +462,38 @@ impl Optimizer {
+/// This differs from [`LogicalPlan::check_invariants`], which addresses if a singular
+/// LogicalPlan is valid. Instead this address if the optimization (before and after)
+/// is valid based upon permitted changes.
+fn assert_valid_optimization(
```
Since all this does is call `assert_expected_schema `I think it might be easier to understand if you changed the callsite to call `assert_expected_schema` directly

wiedld replied:
The purpose of `assert_valid_optimization` is define what is allowed to change in the optimizer passes. As of right now, we only defined that that schema could not change. The idea was that [future work](https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777) could add to this method.

alamb replied:
Make sense

alamb reviewed (commented):
Thank you @wiedld  - I think this is a nice step towards keeping Datafusion stable for building other systems. Other than the performance regression  I think this PR is ready to go.

wiedld commented:
We no longer have any per-optimizer-pass validation in release mode. Instead, the checks run before and after ALL optimizer passes are completed.

When running in debug mode, the per-optimizer-pass will still occur. This will allow users of datafusion to more readily debug when their DF upgrades are failing. 🤞🏼 

Here are the benchmarks. There is no change:
<details>

```
$ critcmp main  check_lp_invariants
group                                         check_lp_invariants                    main
-----                                         -------------------                    ----
logical_aggregate_with_join                   1.00  1001.9±10.39µs        ? ?/sec    1.01  1007.8±10.29µs        ? ?/sec
logical_select_all_from_1000                  1.00      4.8±0.04ms        ? ?/sec    1.00      4.8±0.04ms        ? ?/sec
logical_select_one_from_700                   1.01    753.9±7.37µs        ? ?/sec    1.00    743.5±5.70µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00    716.8±5.68µs        ? ?/sec    1.00    718.2±4.64µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00    699.3±5.84µs        ? ?/sec    1.00    699.6±3.40µs        ? ?/sec
physical_intersection                         1.07   1850.8±9.91µs        ? ?/sec    1.00  1726.6±15.43µs        ? ?/sec
physical_join_consider_sort                   1.17      2.9±0.01ms        ? ?/sec    1.00      2.5±0.01ms        ? ?/sec
physical_join_distinct                        1.00    688.2±5.06µs        ? ?/sec    1.00    689.2±7.02µs        ? ?/sec
physical_many_self_joins                      1.21     16.9±0.12ms        ? ?/sec    1.00     13.9±0.07ms        ? ?/sec
physical_plan_clickbench_all                  1.02    185.0±0.71ms        ? ?/sec    1.00    181.5±0.36ms        ? ?/sec
physical_plan_clickbench_q1                   1.01      2.6±0.02ms        ? ?/sec    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q10                  1.01      3.5±0.01ms        ? ?/sec    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q11                  1.02      3.6±0.02ms        ? ?/sec    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q12                  1.01      3.7±0.01ms        ? ?/sec    1.00      3.7±0.22ms        ? ?/sec
physical_plan_clickbench_q13                  1.02      3.4±0.01ms        ? ?/sec    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q14                  1.01      3.6±0.01ms        ? ?/sec    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q15                  1.02      3.5±0.01ms        ? ?/sec    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q16                  1.01      3.0±0.02ms        ? ?/sec    1.00      2.9±0.01ms        ? ?/sec
physical_plan_clickbench_q17                  1.01      3.1±0.04ms        ? ?/sec    1.00      3.0±0.01ms        ? ?/sec
physical_plan_clickbench_q18                  1.01      2.8±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q19                  1.01      3.5±0.01ms        ? ?/sec    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q2                   1.02      2.8±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q20                  1.02      2.7±0.01ms        ? ?/sec    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q21                  1.02      2.9±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q22                  1.01      3.7±0.01ms        ? ?/sec    1.00      3.6±0.01ms        ? ?/sec
physical_plan_clickbench_q23                  1.01      4.0±0.02ms        ? ?/sec    1.00      4.0±0.02ms        ? ?/sec
physical_plan_clickbench_q24                  1.02      4.8±0.02ms        ? ?/sec    1.00      4.7±0.01ms        ? ?/sec
physical_plan_clickbench_q25                  1.02      3.2±0.02ms        ? ?/sec    1.00      3.1±0.01ms        ? ?/sec
physical_plan_clickbench_q26                  1.02      2.9±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q27                  1.01      3.2±0.01ms        ? ?/sec    1.00      3.2±0.01ms        ? ?/sec
physical_plan_clickbench_q28                  1.01      3.8±0.01ms        ? ?/sec    1.00      3.7±0.02ms        ? ?/sec
physical_plan_clickbench_q29                  1.02      4.8±0.02ms        ? ?/sec    1.00      4.7±0.01ms        ? ?/sec
physical_plan_clickbench_q3                   1.01      2.8±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q30                  1.02     15.6±0.45ms        ? ?/sec    1.00     15.3±0.04ms        ? ?/sec
physical_plan_clickbench_q31                  1.01      3.9±0.02ms        ? ?/sec    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q32                  1.01      3.9±0.02ms        ? ?/sec    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q33                  1.01      3.5±0.01ms        ? ?/sec    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q34                  1.01      3.1±0.01ms        ? ?/sec    1.00      3.1±0.01ms        ? ?/sec
physical_plan_clickbench_q35                  1.01      3.2±0.01ms        ? ?/sec    1.00      3.2±0.01ms        ? ?/sec
physical_plan_clickbench_q36                  1.00      4.3±0.01ms        ? ?/sec    1.00      4.2±0.02ms        ? ?/sec
physical_plan_clickbench_q37                  1.01      4.3±0.01ms        ? ?/sec    1.00      4.3±0.02ms        ? ?/sec
physical_plan_clickbench_q38                  1.01      4.3±0.02ms        ? ?/sec    1.00      4.3±0.01ms        ? ?/sec
physical_plan_clickbench_q39                  1.01      3.9±0.01ms        ? ?/sec    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q4                   1.01      2.6±0.01ms        ? ?/sec    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q40                  1.01      4.4±0.02ms        ? ?/sec    1.00      4.3±0.02ms        ? ?/sec
physical_plan_clickbench_q41                  1.01      4.2±0.01ms        ? ?/sec    1.00      4.1±0.02ms        ? ?/sec
physical_plan_clickbench_q42                  1.02      4.1±0.01ms        ? ?/sec    1.00      4.0±0.01ms        ? ?/sec
physical_plan_clickbench_q43                  1.01      4.1±0.02ms        ? ?/sec    1.00      4.0±0.01ms        ? ?/sec
physical_plan_clickbench_q44                  1.01      2.8±0.01ms        ? ?/sec    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q45                  1.01      2.8±0.01ms        ? ?/sec    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q46                  1.01      3.3±0.01ms        ? ?/sec    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q47                  1.01      3.8±0.01ms        ? ?/sec    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q48                  1.02      4.3±0.02ms        ? ?/sec    1.00      4.3±0.01ms        ? ?/sec
physical_plan_clickbench_q49                  1.02      4.6±0.02ms        ? ?/sec    1.00      4.5±0.01ms        ? ?/sec
physical_plan_clickbench_q5                   1.01      2.8±0.01ms        ? ?/sec    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q6                   1.01      2.8±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q7                   1.01      3.3±0.01ms        ? ?/sec    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q8                   1.01      3.0±0.01ms        ? ?/sec    1.00      3.0±0.01ms        ? ?/sec
physical_plan_clickbench_q9                   1.01      3.3±0.01ms        ? ?/sec    1.00      3.3±0.02ms        ? ?/sec
physical_plan_tpcds_all                       1.01   1178.4±4.95ms        ? ?/sec    1.00   1161.8±5.74ms        ? ?/sec
physical_plan_tpch_all                        1.01     73.3±0.31ms        ? ?/sec    1.00     72.7±0.20ms        ? ?/sec
physical_plan_tpch_q1                         1.00      2.6±0.01ms        ? ?/sec    1.00      2.6±0.01ms        ? ?/sec
physical_plan_tpch_q10                        1.01      3.6±0.01ms        ? ?/sec    1.00      3.5±0.01ms        ? ?/sec
physical_plan_tpch_q11                        1.00      3.1±0.01ms        ? ?/sec    1.01      3.2±0.17ms        ? ?/sec
physical_plan_tpch_q12                        1.00      2.5±0.01ms        ? ?/sec    1.00      2.5±0.01ms        ? ?/sec
physical_plan_tpch_q13                        1.00   1831.4±9.43µs        ? ?/sec    1.00   1832.2±7.35µs        ? ?/sec
physical_plan_tpch_q14                        1.02      2.2±0.01ms        ? ?/sec    1.00      2.2±0.01ms        ? ?/sec
physical_plan_tpch_q16                        1.01      3.2±0.02ms        ? ?/sec    1.00      3.2±0.01ms        ? ?/sec
physical_plan_tpch_q17                        1.01      2.9±0.01ms        ? ?/sec    1.00      2.9±0.01ms        ? ?/sec
physical_plan_tpch_q18                        1.01      3.3±0.01ms        ? ?/sec    1.00      3.2±0.01ms        ? ?/sec
physical_plan_tpch_q19                        1.00      5.3±0.02ms        ? ?/sec    1.00      5.3±0.02ms        ? ?/sec
physical_plan_tpch_q2                         1.00      6.4±0.02ms        ? ?/sec    1.00      6.3±0.02ms        ? ?/sec
physical_plan_tpch_q20                        1.00      3.9±0.01ms        ? ?/sec    1.00      3.8±0.02ms        ? ?/sec
physical_plan_tpch_q21                        1.01      5.1±0.02ms        ? ?/sec    1.00      5.1±0.02ms        ? ?/sec
physical_plan_tpch_q22                        1.01      2.9±0.01ms        ? ?/sec    1.00      2.9±0.01ms        ? ?/sec
physical_plan_tpch_q3                         1.01      2.5±0.01ms        ? ?/sec    1.00      2.5±0.01ms        ? ?/sec
physical_plan_tpch_q4                         1.00      2.0±0.01ms        ? ?/sec    1.00   1993.7±9.04µs        ? ?/sec
physical_plan_tpch_q5                         1.02      3.6±0.14ms        ? ?/sec    1.00      3.5±0.01ms        ? ?/sec
physical_plan_tpch_q6                         1.01  1363.8±14.88µs        ? ?/sec    1.00   1349.8±7.07µs        ? ?/sec
physical_plan_tpch_q7                         1.01      4.7±0.02ms        ? ?/sec    1.00      4.7±0.01ms        ? ?/sec
physical_plan_tpch_q8                         1.01      5.6±0.01ms        ? ?/sec    1.00      5.6±0.02ms        ? ?/sec
physical_plan_tpch_q9                         1.01      4.3±0.01ms        ? ?/sec    1.00      4.3±0.01ms        ? ?/sec
physical_select_aggregates_from_200           1.01     24.4±0.06ms        ? ?/sec    1.00     24.3±0.05ms        ? ?/sec
physical_select_all_from_1000                 1.05     44.8±0.36ms        ? ?/sec    1.00     42.6±0.41ms        ? ?/sec
physical_select_one_from_700                  1.13      2.8±0.02ms        ? ?/sec    1.00      2.5±0.01ms        ? ?/sec
physical_theta_join_consider_sort             1.15      3.2±0.01ms        ? ?/sec    1.00      2.8±0.01ms        ? ?/sec
physical_unnest_to_join                       1.06      2.7±0.06ms        ? ?/sec    1.00      2.5±0.01ms        ? ?/sec
with_param_values_many_columns                1.03    189.2±0.92µs        ? ?/sec    1.00    183.2±0.56µs        ? ?/sec
```
</details>

wiedld commented:
Fixed the performance regression. It wasn't where we thought it was.

The problems was a recursive check (down the LP) of the check_fields within the `assert_unique_field_names()`. I've removed the recursive nature of this check.

<details><summary>Final numbers, no regression</summary>

```
$ critcmp main invariants__no_recursive_namecheck__schema_check_per_optimizer_pass
group                                         invariants__no_recursive_namecheck__schema_check_per_optimizer_pass    main
-----                                         -------------------------------------------------------------------    ----
logical_aggregate_with_join                   1.00    987.0±5.66µs        ? ?/sec                                    1.02  1007.8±10.29µs        ? ?/sec
logical_select_all_from_1000                  1.04      5.0±0.04ms        ? ?/sec                                    1.00      4.8±0.04ms        ? ?/sec
logical_select_one_from_700                   1.00    739.4±5.38µs        ? ?/sec                                    1.01    743.5±5.70µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00    703.2±5.00µs        ? ?/sec                                    1.02    718.2±4.64µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00    685.1±5.28µs        ? ?/sec                                    1.02    699.6±3.40µs        ? ?/sec
physical_intersection                         1.00  1726.0±10.72µs        ? ?/sec                                    1.00  1726.6±15.43µs        ? ?/sec
physical_join_consider_sort                   1.00      2.4±0.02ms        ? ?/sec                                    1.02      2.5±0.01ms        ? ?/sec
physical_join_distinct                        1.00    677.2±5.94µs        ? ?/sec                                    1.02    689.2±7.02µs        ? ?/sec
physical_many_self_joins                      1.00     13.6±0.07ms        ? ?/sec                                    1.02     13.9±0.07ms        ? ?/sec
physical_plan_clickbench_all                  1.00    181.7±0.39ms        ? ?/sec                                    1.00    181.5±0.36ms        ? ?/sec
physical_plan_clickbench_q1                   1.00      2.6±0.01ms        ? ?/sec                                    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q10                  1.00      3.5±0.01ms        ? ?/sec                                    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q11                  1.00      3.5±0.01ms        ? ?/sec                                    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q12                  1.00      3.6±0.01ms        ? ?/sec                                    1.01      3.7±0.22ms        ? ?/sec
physical_plan_clickbench_q13                  1.00      3.3±0.01ms        ? ?/sec                                    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q14                  1.00      3.5±0.01ms        ? ?/sec                                    1.00      3.5±0.01ms        ? ?/sec
physical_plan_clickbench_q15                  1.00      3.4±0.01ms        ? ?/sec                                    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q16                  1.00      3.0±0.01ms        ? ?/sec                                    1.00      2.9±0.01ms        ? ?/sec
physical_plan_clickbench_q17                  1.00      3.0±0.01ms        ? ?/sec                                    1.00      3.0±0.01ms        ? ?/sec
physical_plan_clickbench_q18                  1.01      2.8±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q19                  1.00      3.5±0.01ms        ? ?/sec                                    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q2                   1.00      2.8±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q20                  1.00      2.6±0.01ms        ? ?/sec                                    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q21                  1.01      2.8±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q22                  1.00      3.6±0.01ms        ? ?/sec                                    1.01      3.6±0.01ms        ? ?/sec
physical_plan_clickbench_q23                  1.00      4.0±0.01ms        ? ?/sec                                    1.00      4.0±0.02ms        ? ?/sec
physical_plan_clickbench_q24                  1.01      4.8±0.02ms        ? ?/sec                                    1.00      4.7±0.01ms        ? ?/sec
physical_plan_clickbench_q25                  1.01      3.2±0.02ms        ? ?/sec                                    1.00      3.1±0.01ms        ? ?/sec
physical_plan_clickbench_q26                  1.01      2.9±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q27                  1.00      3.2±0.01ms        ? ?/sec                                    1.00      3.2±0.01ms        ? ?/sec
physical_plan_clickbench_q28                  1.00      3.7±0.01ms        ? ?/sec                                    1.00      3.7±0.02ms        ? ?/sec
physical_plan_clickbench_q29                  1.00      4.7±0.02ms        ? ?/sec                                    1.00      4.7±0.01ms        ? ?/sec
physical_plan_clickbench_q3                   1.01      2.8±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q30                  1.01     15.5±0.41ms        ? ?/sec                                    1.00     15.3±0.04ms        ? ?/sec
physical_plan_clickbench_q31                  1.00      3.8±0.02ms        ? ?/sec                                    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q32                  1.00      3.8±0.01ms        ? ?/sec                                    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q33                  1.00      3.4±0.01ms        ? ?/sec                                    1.00      3.4±0.01ms        ? ?/sec
physical_plan_clickbench_q34                  1.00      3.1±0.01ms        ? ?/sec                                    1.00      3.1±0.01ms        ? ?/sec
physical_plan_clickbench_q35                  1.00      3.2±0.01ms        ? ?/sec                                    1.00      3.2±0.01ms        ? ?/sec
physical_plan_clickbench_q36                  1.00      4.2±0.01ms        ? ?/sec                                    1.00      4.2±0.02ms        ? ?/sec
physical_plan_clickbench_q37                  1.00      4.3±0.01ms        ? ?/sec                                    1.00      4.3±0.02ms        ? ?/sec
physical_plan_clickbench_q38                  1.00      4.3±0.02ms        ? ?/sec                                    1.00      4.3±0.01ms        ? ?/sec
physical_plan_clickbench_q39                  1.00      3.8±0.01ms        ? ?/sec                                    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q4                   1.00      2.6±0.01ms        ? ?/sec                                    1.00      2.6±0.01ms        ? ?/sec
physical_plan_clickbench_q40                  1.00      4.3±0.01ms        ? ?/sec                                    1.00      4.3±0.02ms        ? ?/sec
physical_plan_clickbench_q41                  1.00      4.1±0.01ms        ? ?/sec                                    1.00      4.1±0.02ms        ? ?/sec
physical_plan_clickbench_q42                  1.00      4.0±0.02ms        ? ?/sec                                    1.00      4.0±0.01ms        ? ?/sec
physical_plan_clickbench_q43                  1.00      4.0±0.01ms        ? ?/sec                                    1.00      4.0±0.01ms        ? ?/sec
physical_plan_clickbench_q44                  1.00      2.7±0.01ms        ? ?/sec                                    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q45                  1.00      2.8±0.01ms        ? ?/sec                                    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q46                  1.00      3.2±0.01ms        ? ?/sec                                    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q47                  1.00      3.8±0.01ms        ? ?/sec                                    1.00      3.8±0.01ms        ? ?/sec
physical_plan_clickbench_q48                  1.00      4.3±0.01ms        ? ?/sec                                    1.00      4.3±0.01ms        ? ?/sec
physical_plan_clickbench_q49                  1.00      4.5±0.02ms        ? ?/sec                                    1.00      4.5±0.01ms        ? ?/sec
physical_plan_clickbench_q5                   1.00      2.8±0.01ms        ? ?/sec                                    1.00      2.7±0.01ms        ? ?/sec
physical_plan_clickbench_q6                   1.00      2.8±0.01ms        ? ?/sec                                    1.00      2.8±0.01ms        ? ?/sec
physical_plan_clickbench_q7                   1.01      3.3±0.01ms        ? ?/sec                                    1.00      3.3±0.01ms        ? ?/sec
physical_plan_clickbench_q8                   1.00      3.0±0.01ms        ? ?/sec                                    1.00      3.0±0.01ms        ? ?/sec
physical_plan_clickbench_q9                   1.00      3.2±0.02ms        ? ?/sec                                    1.00      3.3±0.02ms        ? ?/sec
physical_plan_tpcds_all                       1.00   1160.9±2.22ms        ? ?/sec                                    1.00   1161.8±5.74ms        ? ?/sec
physical_plan_tpch_all                        1.00     72.5±0.23ms        ? ?/sec                                    1.00     72.7±0.20ms        ? ?/sec
physical_plan_tpch_q1                         1.00      2.6±0.01ms        ? ?/sec                                    1.00      2.6±0.01ms        ? ?/sec
physical_plan_tpch_q10                        1.00      3.5±0.01ms        ? ?/sec                                    1.00      3.5±0.01ms        ? ?/sec
physical_plan_tpch_q11                        1.00      3.1±0.01ms        ? ?/sec                                    1.01      3.2±0.17ms        ? ?/sec
physical_plan_tpch_q12                        1.00      2.5±0.01ms        ? ?/sec                                    1.00      2.5±0.01ms        ? ?/sec
physical_plan_tpch_q13                        1.00   1832.8±8.27µs        ? ?/sec                                    1.00   1832.2±7.35µs        ? ?/sec
physical_plan_tpch_q14                        1.00      2.2±0.01ms        ? ?/sec                                    1.00      2.2±0.01ms        ? ?/sec
physical_plan_tpch_q16                        1.00      3.2±0.01ms        ? ?/sec                                    1.00      3.2±0.01ms        ? ?/sec
physical_plan_tpch_q17                        1.00      2.9±0.01ms        ? ?/sec                                    1.00      2.9±0.01ms        ? ?/sec
physical_plan_tpch_q18                        1.00      3.2±0.01ms        ? ?/sec                                    1.00      3.2±0.01ms        ? ?/sec
physical_plan_tpch_q19                        1.00      5.3±0.01ms        ? ?/sec                                    1.00      5.3±0.02ms        ? ?/sec
physical_plan_tpch_q2                         1.01      6.4±0.03ms        ? ?/sec                                    1.00      6.3±0.02ms        ? ?/sec
physical_plan_tpch_q20                        1.00      3.9±0.01ms        ? ?/sec                                    1.00      3.8±0.02ms        ? ?/sec
physical_plan_tpch_q21                        1.00      5.1±0.01ms        ? ?/sec                                    1.00      5.1±0.02ms        ? ?/sec
physical_plan_tpch_q22                        1.00      2.9±0.01ms        ? ?/sec                                    1.00      2.9±0.01ms        ? ?/sec
physical_plan_tpch_q3                         1.00      2.5±0.01ms        ? ?/sec                                    1.00      2.5±0.01ms        ? ?/sec
physical_plan_tpch_q4                         1.00   1994.6±9.01µs        ? ?/sec                                    1.00   1993.7±9.04µs        ? ?/sec
physical_plan_tpch_q5                         1.01      3.6±0.02ms        ? ?/sec                                    1.00      3.5±0.01ms        ? ?/sec
physical_plan_tpch_q6                         1.00   1350.1±7.99µs        ? ?/sec                                    1.00   1349.8±7.07µs        ? ?/sec
physical_plan_tpch_q7                         1.01      4.7±0.02ms        ? ?/sec                                    1.00      4.7±0.01ms        ? ?/sec
physical_plan_tpch_q8                         1.01      5.6±0.03ms        ? ?/sec                                    1.00      5.6±0.02ms        ? ?/sec
physical_plan_tpch_q9                         1.00      4.3±0.02ms        ? ?/sec                                    1.00      4.3±0.01ms        ? ?/sec
physical_select_aggregates_from_200           1.00     24.3±0.09ms        ? ?/sec                                    1.00     24.3±0.05ms        ? ?/sec
physical_select_all_from_1000                 1.01     43.2±0.35ms        ? ?/sec                                    1.00     42.6±0.41ms        ? ?/sec
physical_select_one_from_700                  1.00      2.5±0.01ms        ? ?/sec                                    1.02      2.5±0.01ms        ? ?/sec
physical_theta_join_consider_sort             1.00      2.7±0.02ms        ? ?/sec                                    1.02      2.8±0.01ms        ? ?/sec
physical_unnest_to_join                       1.00      2.4±0.01ms        ? ?/sec                                    1.02      2.5±0.01ms        ? ?/sec
with_param_values_many_columns                1.00    183.9±0.62µs        ? ?/sec                                    1.00    183.2±0.56µs        ? ?/sec
```
</details>

The release vs debug mode only has a single difference in the checks. The debug mode will run a full `InvariantLevel::Executable check` per each optimizer pass (instead of after all pass).

alamb commented on datafusion/optimizer/src/analyzer/mod.rs:
```diff
@@ -140,6 +143,14 @@ impl Analyzer {
+        // verify the logical plan required invariants at the start, before analyzer
+        plan.check_invariants(InvariantLevel::Always).map_err(|e| {
+            DataFusionError::Context(
+                "assert_lp_invariants_before_analyzers".to_string(),
```
I also think a more human readable message might be nice, for example instead of 

"assert_lp_invariants_before_analyzers"

Maybe something like 

"Invalid input plan passed to Analyzer"

alamb reviewed (approved):
Thank you @wiedld  -- this looks like a great improvement

I left some comments but I don't they are required

In my opinion this is needed before merging:
1. I will wait 24 to allow time for anyone else to comment
2. I am also double-checking  the benchmarks

If you have time I do think some of my comments would improve the usability of this PR as well (e.g. the message changes)

Also, shall we file some follow on tickets for follow on work?

berkaysynnada commented on datafusion/expr/src/logical_plan/invariants.rs:
```diff
@@ -15,14 +15,98 @@
+    /// such as the type and number of function arguments are correct and
+    /// that wildcards have been expanded
+    ///
+    /// To ensure a LogicalPlan satisfies the `Executable` inariants, run the
```
```suggestion
    /// To ensure a LogicalPlan satisfies the `Executable` invariants, run the
```

berkaysynnada commented on datafusion/optimizer/src/optimizer.rs:
```diff
@@ -384,9 +396,26 @@ impl Optimizer {
-                // verify the rule didn't change the schema
                 .and_then(|tnr| {
-                    assert_schema_is_the_same(rule.name(), &starting_schema, &tnr.data)?;
+                    // run checks optimizer invariant checks, per pass
```
what did you mean exactly here?

wiedld replied:
The correct statement is `// run checks optimizer invariant checks, per optimizer rule applied`. And then multiple rules are applied per pass. Statement is fixed. Thanks!.

alamb commented:
My performance benchmarks show also show now difference ✅ 

```
++ critcmp main 13525_invariant-checking-for-implicit-LP-changes
group                                         13525_invariant-checking-for-implicit-LP-changes    main
-----                                         ------------------------------------------------    ----
logical_aggregate_with_join                   1.00  1498.7±38.07µs        ? ?/sec                 1.00  1494.0±18.89µs        ? ?/sec
logical_select_all_from_1000                  1.00      5.3±0.05ms        ? ?/sec                 1.00      5.3±0.05ms        ? ?/sec
logical_select_one_from_700                   1.00  1169.9±26.47µs        ? ?/sec                 1.01  1179.5±14.41µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00  1154.9±13.48µs        ? ?/sec                 1.00  1158.7±16.27µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00  1140.9±18.36µs        ? ?/sec                 1.00  1140.3±14.97µs        ? ?/sec
physical_intersection                         1.00      2.5±0.02ms        ? ?/sec                 1.00      2.5±0.02ms        ? ?/sec
physical_join_consider_sort                   1.00      3.3±0.02ms        ? ?/sec                 1.00      3.3±0.04ms        ? ?/sec
physical_join_distinct                        1.00  1130.5±14.83µs        ? ?/sec                 1.00  1133.2±15.97µs        ? ?/sec
physical_many_self_joins                      1.01     17.5±0.10ms        ? ?/sec                 1.00     17.4±0.08ms        ? ?/sec
physical_plan_clickbench_all                  1.00    227.8±1.70ms        ? ?/sec                 1.01    230.0±3.38ms        ? ?/sec

```

jonahgao reviewed (approved):
LGTM👍, thanks @wiedld

wiedld commented:
Thank you for the reviews. 
I made the requested changes to the terminology and error messages.

.

> Also, shall we file some follow on tickets for follow on work?

I have a task lists here for [follow up items](https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777). I was planning to dig more into code to assess viability (& I can make tickets then too) as I start implementing.

wiedld commented on datafusion/sqllogictest/test_files/subquery.slt:
```diff
@@ -433,17 +433,32 @@ logical_plan
 
 #invalid_scalar_subquery
-statement error DataFusion error: check_analyzed_plan\ncaused by\nError during planning: Scalar subquery should only return one column, but found 2: t2.t2_id, t2.t2_name
+statement error
 SELECT t1_id, t1_name, t1_int, (select t2_id, t2_name FROM t2 WHERE t2.t2_id = t1.t1_int) FROM t1
+----
+DataFusion error: Invalid (non-executable) plan after Analyzer
+caused by
+Error during planning: Scalar subquery should only return one column, but found 2: t2.t2_id, t2.t2_name
```
This change is was introduced with `cargo test --test sqllogictests -- --complete`, passing locally on macos -- but fails in CI. I'll figure out the whys later.

jonahgao replied:
sqllogictest's behavior with multi-line errors is strange. I haven't debugged it, but using single-line errors should be able to fix it.
```
statement error DataFusion error: Invalid \(non-executable\) plan after Analyzer\ncaused by\nError during planning: Scalar subquery should only return one column, but found 2: t2.t2_id, t2.t2_name
SELECT t1_id, t1_name, t1_int, (select t2_id, t2_name FROM t2 WHERE t2.t2_id = t1.t1_int) FROM t1
```

wiedld replied:
Got it. Fixed. Thanks @jonahgao .

alamb commented:
Looks like Ci is failint a bit here

alamb reviewed (approved):
Thank you @wiedld  -- I think this is a nice step forward in helping avoid bugs in DataFusion ❤️ 

Thanks @jonahgao  @findepi @Sl1mb0  and @berkaysynnada 

I'll merge this PR now and we can address any other suggestions as a follow on PR
